tls/s2n_handshake_io.c

```diff
@@ -323,7 +325,12 @@ int s2n_conn_set_handshake_type(struct s2n_connection *conn)
 
     s2n_cert_auth_type client_cert_auth_type;
     GUARD(s2n_connection_get_client_auth_type(conn, &client_cert_auth_type));
-    if(client_cert_auth_type != S2N_CERT_AUTH_NONE) {
+
+    if (conn->mode == S2N_CLIENT && client_cert_auth_type == S2N_CERT_AUTH_REQUIRED) {
+        /* If we're a client, and Client Auth is REQUIRED, then the Client must expect the CLIENT_CERT_REQ Message */
+        conn->handshake.handshake_type |= CLIENT_AUTH;
```

Comment:
Since the client is dynamically reacting to the presence of a CertificateRequest now, do we still need to have the client explicitly setting a client auth mode?

Reply:
We're still honoring the `client_cert_auth_type` value. 

If it's `NONE` we'll abort the handshake rather than enter the ClientCertRequest state, if `OPTIONAL` we'll enter that state only if the Server asks and potentially send an empty response if no ClientCert is configured, and if `REQUIRED` we'll always expect the ClientCertRequest from the server and abort the handshake if no ClientCert is configured.

Reply:
Sorry, I think my comment was a little confusing. I meant to ask if we could relax the constraint of `REQUIRED` altogether. Could the client just dynamically react to the CertificateRequest in all cases (`REQUIRED`, `OPTIONAL`, and `NONE`)? If it has a client cert configured and it receives a CertificateRequest, send the cert. If it does not have a client cert configured and it receives a CertificateRequest, send an empty response. If no CertificateRequest is sent, carry on.

If the client is configured to `REQUIRED` and the server does not send the CertificateRequest, why fail the handshake? Why not just let the server decide the requirements of client authentication? Maybe I'm missing a use case or complexity here

Reply:
In that case, wouldn't the client still have authenticated the server to ensure it was communicating with a trusted endpoint? The server would have still sent it's own certificate. The client could enforce server auth without having to enforce mutual auth.

Reply:
The idea of optional client certs doesn't map to the client very well. The RFC says:

> If no suitable certificate is available, the client MUST send a certificate message containing no certificates.
> ...
> If the client does not send any certificates, the server MAY at its discretion either continue the handshake without client authentication, or respond with a fatal handshake_failure alert.

I interpret that as basically saying the RFC doesn't even really allow for clients deciding to end the handshake based on the existence (or lack of) a certificate request. The rules for either party of a handshake sending an error are a little fuzzy in that you _could_ interpret it to mean that if an implementation decides something is an error then it can bail-out, but I interpret it more that if an error as defined by the RFC is encountered then bail out. Since the client certificate portion doesn't allow the client to determine that there's been an error, it wouldn't be valid for the client to bail-out here.

All that is to say that per the RFC, I think setting any of the `REQUIRED/OPTIONAL/NONE` in a client context is invalid. If the consumer, at the application level, wants to require the use of a client cert then `s2n_connection_client_cert_used()` is the mechanism for allowing that (after the handshake is complete).

If you want a feature that allows the client to bail out of connections based on the server's client certificate behavior, I think there's probably a few use-cases for that and it wouldn't be a big deal in practice, but I don't think it would be compliant with the RFC.

Reply:
Opened https://github.com/awslabs/s2n/issues/819 to continue this discussion.